### PR TITLE
Enable prerelease version tags, bump to v1.9.1-prerelease

### PR DESCRIPTION
**Pull request by colin-higgins**
v1.9.1-prerelease

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by bobuva** [comment]
Before we merge this, let's manually run MSI tests including upgrading to make sure it works as expected.

**Comment by colin-higgins** [comment]
@bobuva The only thing that changes in this PR is the nuget packaging. I'll test the MSI after making the versioning changes we talked about, for basing the MSI verison on something hard coded.

**Comment by zacharycmontoya** [comment]
@colin-higgins Sorry, I wasn't clear if this PR would address the MSI versioning, but if you were going to tackle that, you could either hard-code the version directly in `Product.wxs`, or hard-code the `InstallerVersion` property that is set in `Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj`, that way the command line argument is no longer used.

**Comment by zacharycmontoya** [comment]
Maybe it would just be easier to set the values in `Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj` because that is where you can set the `OutputName` property to generate the filename with the prerelease tag as part of the filename. Just be sure to delete the `InstallerVersion` and `OutputName` command line arguments in the `packages.yml` file

**Comment by zacharycmontoya** [comment]
And yet another suggestion, maybe just remove the `Condition="'$(InstallerVersion)'==''"` and `Condition="'$(OutputName)'==''" that way we never override the version from command line arguments, we always rely on the definition in the wixproj file

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj [hunk 1]
```diff
@@ -14,25 +14,18 @@
     <IntermediateOutputPath>obj\$(Configuration)\$(Platform)\</IntermediateOutputPath>
     <SuppressPdbOutput>True</SuppressPdbOutput>
     <DefineSolutionProperties>false</DefineSolutionProperties>
-    <!-- Custom InstallerVersion property for the generated installer,
-         value is set to build constant InstallerVersion below and
-         accessed as $(var.InstallerVersion) from WiX files.
-         This is the default value if one is not specified when building. -->
-    <InstallerVersion Condition="'$(InstallerVersion)'==''">0.1</InstallerVersion>
+    <InstallerVersion>3.9.8</InstallerVersion>
```
Comment (lucaspimentel) [inline deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: Should this be `1.9.1`?

File: tools/Datadog.Core.Tools/EnvironmentTools.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /// General use utility methods for all tests and tools.
+    /// </summary>
+    public class EnvironmentTools
+    {
+        private static string _solutionDirectory = null;
+
+        /// <summary>
+        /// Find the solution directory from anywhere in the hierarchy.
+        /// </summary>
+        public static string GetSolutionDirectory()
```
Comment (colin-higgins) [inline tools/Datadog.Core.Tools/EnvironmentTools.cs]: This seems like a better place for helper utilities. I'll expand on it later and convert uses of EnvironmentHelper.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj [hunk 3]
```diff
@@ -14,25 +14,17 @@
...
+    <InstallerVersion>1.9.1</InstallerVersion>
+    <OutputName>datadog-dotnet-apm-1.9.1-prerelease-$(Platform)-$(Configuration)</OutputName>
   </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|x86' ">
-    <DefineConstants>Debug;InstallerVersion=$(InstallerVersion)</DefineConstants>
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|x86' ">
-    <DefineConstants>InstallerVersion=$(InstallerVersion)</DefineConstants>
+  <PropertyGroup>
+    <DefineConstants>InstallerVersion=1.9.1</DefineConstants>
```
Comment (lucaspimentel) [inline deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: Do we need this? We can already read the `<InstallerVersion>` property above in the wxs file.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Product.wxs [hunk 4]
```diff
@@ -6,7 +6,7 @@
   <Product Id="*"
            Name="$(var.ProductName)"
            Language="1033"
-           Version="$(var.InstallerVersion)"
+           Version="$(InstallerVersion)"
```
Comment (lucaspimentel) [inline deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Product.wxs]: We need the `var.` prefix to read build properties.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj (hunk 3 above)
Comment (zacharycmontoya) [reply deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: Might be worth testing, but from first appearances, this is required to pass the value so the wxs file can access it from `$(var.InstallerVersion)`

Comment (zacharycmontoya) [reply deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: One separate optimization could be to set this to `InstallerVersion=$(InstallerVersion)` that way you don't have to manipulate this line at all, you'd only have to manipulate the above lines.

File: tools/Datadog.Core.Tools/Datadog.Core.Tools.csproj [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    <TargetFrameworks>net452;net461;netcoreapp2.1;netcoreapp3.0</TargetFrameworks>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Linq" Version="4.3.0" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="System.IO.FileSystem">
+      <HintPath>..\..\..\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\3.0.0\ref\netcoreapp3.0\System.IO.FileSystem.dll</HintPath>
```
Comment (lucaspimentel) [inline tools/Datadog.Core.Tools/Datadog.Core.Tools.csproj]: This should be a nuget package reference, not an assembly reference.

Comment (colin-higgins) [reply tools/Datadog.Core.Tools/Datadog.Core.Tools.csproj]: Not sure how that even happened, thanks resharper.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj (hunk 3 above)
Comment (colin-higgins) [reply deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: Once it's set it's set and I don't care all too much, can we leave it in?

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Product.wxs (hunk 4 above)
Comment (lucaspimentel) [inline deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Product.wxs]: This will result in a syntax error from the compiler.

Comment (colin-higgins) [reply deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Product.wxs]: Yep, I noticed. It's fixed.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj (hunk 3 above)
Comment (lucaspimentel) [reply deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: This is the right one. It's the build property `<InstallerVersion>` that we don't need anymore. It was there so we could set the filename. I'll add a suggestion so you only need to click accept.

File: deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj [hunk 6]
```diff
@@ -14,25 +14,14 @@
...
     <SuppressPdbOutput>True</SuppressPdbOutput>
     <DefineSolutionProperties>false</DefineSolutionProperties>
-    <!-- Custom InstallerVersion property for the generated installer,
-         value is set to build constant InstallerVersion below and
-         accessed as $(var.InstallerVersion) from WiX files.
-         This is the default value if one is not specified when building. -->
-    <InstallerVersion Condition="'$(InstallerVersion)'==''">0.1</InstallerVersion>
-    <!-- default msi filename if not specified -->
-    <OutputName Condition="'$(OutputName)'==''">datadog-dotnet-apm-$(InstallerVersion)-$(Platform)-$(Configuration)</OutputName>
+    <InstallerVersion>1.9.1</InstallerVersion>
```
Comment (lucaspimentel) [inline deploy/Datadog.Trace.ClrProfiler.WindowsInstaller/Datadog.Trace.ClrProfiler.WindowsInstaller.wixproj]: We don't need this anymore. It was used to set both the filename and to define the constant, but those are both hard-coded now.
```suggestion
```

File: tools/Datadog.Core.Tools/Datadog.Core.Tools.csproj [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>net452;net461;netcoreapp2.1;netcoreapp3.0</TargetFrameworks>
```
Comment (zacharycmontoya) [inline tools/Datadog.Core.Tools/Datadog.Core.Tools.csproj]: nit: These can probably be reduced to something like `net45;netstandard2.0` but no need to block on this.

File: .azure-pipelines/packages.yml [hunk 8]
```diff
@@ -83,7 +83,7 @@ jobs:
       solution: Datadog.Trace.proj
       platform: $(buildPlatform)
       configuration: $(buildConfiguration)
-      msbuildArguments: /t:msi /p:InstallerVersion=%GitVersion_MajorMinorPatch%;OutputName=datadog-dotnet-apm-%GitVersion_MajorMinorPatch%-$(buildPlatform);RunWixToolsOutOfProc=true
+      msbuildArguments: /t:msi /p:RunWixToolsOutOfProc=true
       maximumCpuCount: true
 
   - task: PublishPipelineArtifact@0
```
Comment (zacharycmontoya) [inline .azure-pipelines/packages.yml]: The `targetPath` below will need updating. Can you update with the relative directory you saw in local testing?

**Comment by colin-higgins** [comment]
Just need to update the release docs... :D

**Pull request merged**